### PR TITLE
Cache deposit signature verification

### PR DESCRIPTION
**Pull request by paulhauner**
## Issue Addressed

NA

## Proposed Changes

- Checks the signature of each eth1 deposit as it comes into our cache, stores the result.
- Replaces a binary search in `DepositCache` since it's there's no guarantee it would always find the log with the highest index for the searched block. New solution is O(n) instead of O(log(n)), but I think we can address this if it becomes an issue during block production.

## Reasoning

This fixes an issue I had observed where when we wait for genesis where:

- `MIN_GENESIS_TIME` has been satisfied.
- There are `MIN_GENESIS_ACTIVE_VALIDATOR_COUNT` deposits, but enough of them are invalid that we drop below that quota.

In this case, we verify all deposit signatures and generate a `BeaconState` for every eth1 block we get. This gets very slow.

## TODO

- [x] ~~Use cached signature verification value when producing genesis states.~~ We can do this in a future PR, if it turns out to be useful. For now it requires changes to `state_processing` and we're on a timeline that's not great for fiddling with consensus code.

## Notes

- ~~Blocked on #1295~~

### CODE REVIEW

File: beacon_node/genesis/src/eth1_genesis_service.rs [hunk 1]
```diff
@@ -287,6 +291,21 @@ impl Eth1GenesisService {
                 continue;
             }
 
+            let valid_signature_count = eth1_service
+                .get_valid_signature_count_at_block(block.number)
+                .unwrap_or(0);
+            if (valid_signature_count as u64) < spec.min_genesis_active_validator_count {
```
Comment (michaelsproul) [inline beacon_node/genesis/src/eth1_genesis_service.rs:297]: Unnecessary parens

File: beacon_node/genesis/src/eth1_genesis_service.rs [hunk 2]
```diff
@@ -287,6 +291,21 @@ impl Eth1GenesisService {
...
+                trace!(
+                    log,
+                    "Insufficient valid signatures";
+                    "genesis_delay" => spec.genesis_delay,
+                    "valid_siganture_count" => valid_signature_count,
```
Comment (michaelsproul) [inline beacon_node/genesis/src/eth1_genesis_service.rs:302]: typo: siganture -> signature

File: beacon_node/genesis/src/eth1_genesis_service.rs (hunk 1 above)
Comment (paulhauner) [reply beacon_node/genesis/src/eth1_genesis_service.rs:297]:
```
error: `<` is interpreted as a start of generic arguments for `u64`, not a comparison
   --> beacon_node/genesis/src/eth1_genesis_service.rs:297:45
    |
297 |             if valid_signature_count as u64 < spec.min_genesis_active_validator_count {
    |                ---------------------------- ^ ----- interpreted as generic arguments
    |                |                            |
    |                |                            not interpreted as comparison
    |                help: try comparing the cast value: `(valid_signature_count as u64)`
```

**Comment by paulhauner** [comment]
All comments addressed @michaelsproul :)

File: beacon_node/genesis/src/eth1_genesis_service.rs (hunk 1 above)
Comment (michaelsproul) [reply beacon_node/genesis/src/eth1_genesis_service.rs:297]: Ewwww

Ok haha

**Pull request merged**